### ISSUE TITLE
Imap4 - issuing a typo causes unhandled exception

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4000]
|<img alt="Magneto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| Magneto reported|
|-|-|
|Trac ID|trac#4000|
|Type|defect|
|Created|2009-09-04 12:24:56Z|

Problem lies in this code (line 5131)
        state = self.state.pop()

If you issue this command for e.g "fetch 1 fulll" (note the 3rd l) 
You will get :
       exceptions.IndexError: pop from empty list
And the client telnet screen gets a :
       3 BAD Server error: pop from empty list

This is because the "l" is treated as a command but the "self.state" is already empty.

Suggested fix : 
                if not self.state :
                    self.remaining = ""
                    break
                

Attachments:

* [4000-IMAP4FetchParser-20120704.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e0b/e0b2d81c80b94799aeef126f990f48cdb5505bbc/a8eee94dd4de67a73e7d45d0c08d8686a87bf713.patch) (1494 bytes) - added by argonemyth on 2012-07-04 08:35:11Z - 
* [4000-IMAP4FetchParser-20120724.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e0b/e0b2d81c80b94799aeef126f990f48cdb5505bbc/9b8c2015eb771a3918538dee5455afa31c74255b.patch) (2078 bytes) - added by argonemyth on 2012-07-24 14:43:47Z - 
* [4000-IMAP4FetchParser-20120807.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/e0b/e0b2d81c80b94799aeef126f990f48cdb5505bbc/55bab467029ee0a88dee7b09651228824806c188.patch) (2421 bytes) - added by argonemyth on 2012-08-07 13:16:06Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4000 4000
type__defect defect
reporter__Magneto Magneto
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__imap4_fetchparser imap4 fetchparser
time__1252067096000000 1252067096000000
changetime__1348365326000000 1348365326000000
version__None None
owner__ 
cc__argonemyth
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Thanks for reporting this.  It would be good to have it fixed.  Do you think you can supply a test case and a patch?  That would help us a lot. :)

As this isn't a regression since 8.2, I'm going to drop it from the 9.0 milestone.  That doesn't mean we don't want to fix it, just that it isn't a release blocker for 9.0.

**Comment by twisted-trac** [comment]
|<img alt="Magneto's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|Magneto set owner to Magneto<br>Magneto set status to `assigned`|
|-|-|

Hi,
yes i can fix it and many others, but i think there are a lot more so maybe i should issue a ONE big DIFF and unit test once it can handle all my current test. I am testing against various clients from mobile to webmails. 

I have just configured the trac here to send to my email (i can be reached here) marcus@.../marcus@... 
- sorry for the late reply, couldn't find this in the trac->my tickets. 

Temporal fix :
    if not self.state :
         self.remaining = ""
         break

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_3"></a>|@argonemyth set owner to @argonemyth<br>@argonemyth set status to `new`|
|-|-|

I will make a patch for the fix.

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_4"></a>|@argonemyth commented|
|-|-|

Please review attached patch for the typo issue. I put s= "" instead of the suggested self.remaining = "".

Also, a simple unit test is added in **twisted.mail.test.test_imap.IMAP4HelperTestCase**.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @argonemyth|
|-|-|

Thanks.

This change seems to make the parser accept any extra bytes at the end of a valid FETCH command?  The current behavior, to send a BAD response, is correct according to the RFC.  It's not clear why the parser should be changed this way.

What is clear is that there should be unit tests covering the invalid argument case.  But I think they should demonstrate that a BAD response is sent, not that the command is luckily parsed as what the client hopefully meant.

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_7"></a>|@argonemyth commented|
|-|-|

Replying to [exarkun](#note_6):

> This change seems to make the parser accept any extra bytes at the end of a valid FETCH command?  The current behavior, to send a BAD response, is correct according to the RFC.  It's not clear why the parser should be changed this way.
>
I agree with you! Sending a BAD response should be the correct way of handling an invalid command/argument (regardless it's a typo or other reasons). But why did you encourage Magneto to fix the problem his way? I merely did what you asked for 3 years ago. According to the spec, we should return a **BAD arguments invalid** response instead of the 'server error' response, I will modify the patch to do that.

> What is clear is that there should be unit tests covering the invalid argument case.  But I think they should demonstrate that a BAD response is sent, not that the command is luckily parsed as what the client hopefully meant.
> 

I will also modify the test case to expect a BAD response.

Thanks for the review! :)

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_8"></a>|@argonemyth removed owner|
|-|-|

A new patch has been made:

1. I think manually raise an exception here is better than the client getting an **BAD Server error**, as 'pop from empty list' is difficult to make sense of. Plus, according to RFC3501, it's better to indicate that it's an invalid argument.

2. I added a new test case in the end - we need a IMAP4 server that uses StringTransport() for some fetch related tests. Right now, it's only tailored for this specific fetch test, but could later be expended for other tests.

Please review the patch & let me know if you want me to change anything. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @argonemyth|
|-|-|

Thanks!

  1. I think you can get away without setting the `mbox` attribute in this test.  The fetch command doesn't actually get evaluated, and `self` (the value being set) is not actually a mailbox in this case.
  1. A news fragment would be excellent :)

Otherwise, this looks great.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="argonemyth's avatar" src="https://avatars.githubusercontent.com/u/965739?s=50" width="50" height="50">](https://github.com/argonemyth)<a name="note_10"></a>|@argonemyth removed owner|
|-|-|

Replying to [exarkun](#note_9):

A new patch was uploaded!

> 
>   1. I think you can get away without setting the `mbox` attribute in this test.  The fetch command doesn't actually get evaluated, and `self` (the value being set) is not actually a mailbox in this case.

You are right, I removed the attribute.

>   2. A news fragment would be excellent :)

Added a new file **4000.bugfix**.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_11"></a>|@jerub commented|
|-|-|

Thanks for the patch! It looks good.

**Comment by twisted-trac** [comment]
|[<img alt="jerub's avatar" src="https://avatars.githubusercontent.com/u/315461?s=50" width="50" height="50">](https://github.com/jerub)<a name="note_12"></a>|@jerub set status to `closed`|
|-|-|

(In [35782]) Apply 4000-IMAP4FetchParser-20120807.patch: raise IllegalClientResponse when extra data is sent at the end of a FETCH

Author: Magneto
Reviewer: jerub
Fixes: [#4000](https://github.com/twisted/twisted/issues/4000)
